Conformance profiles modes and versions (#2391)
kubernetes-sigs/gateway-api: geps/gep-1709.md

[go]:https://go.dev
[lib]:https://pkg.go.dev/sigs.k8s.io/gateway-api@v0.6.2/conformance/utils/suite

### Implementation mode

The certification process runs against an implementation using a specific mode
(e.g., `standard`, `enhanced`) that is defaulted to `standard`. The final

> report will contain such a mode in the modes field. Every certification report
> `modes` array contains only one element. Multiple reports can be merged into an

Reviewer: This patch tries to add more definition to what we mean by "mode".

```suggestion
report will contain such a mode in the modes field. A "mode" is intended to capture
situations where a Gateway API implementation may have different features and capabilities
depending on how it is deployed (e.g. an implementation might be deployed with an external,
OR internal load balancer, and have different capabilities depending on the mode chosen).
Every certification report `modes` array contains only one element. Reports can be merged
into one another to provide a single report that accounts for multiple modes, given the same
Gateway API version.
```

omni-comprehensive one, as described [later](#multiple-reports-merge).

### Implementation version

Every report can specify multiple results, dependent on the tested version of the

> certifying implementation. Each certification run uses a specific version of the

Reviewer: ```suggestion
implementation being certified. Each certification run uses a specific version of the
```

> implementation, multiple certification processes can be merged as explained [later](#multiple-reports-merge).

Reviewer: ```suggestion
implementation, multiple certification processes can be merged as explained [below](#multiple-reports-merge).
```

[go]:https://go.dev
[lib]:https://pkg.go.dev/sigs.k8s.io/gateway-api@v0.6.2/conformance/utils/suite

### Implementation mode

The certification process runs against an implementation using a specific mode
(e.g., `standard`, `enhanced`) that is defaulted to `standard`. The final
report will contain such a mode in the modes field. A "mode" is intended to capture
situations where a Gateway API implementation may have different features and capabilities
depending on how it is deployed (e.g. an implementation might be deployed with an external,
OR internal load balancer, and have different capabilities depending on the mode chosen).
Every certification report `modes` array contains only one element. Reports can be merged
into one another to provide a single report that accounts for multiple modes, given the same

> Gateway API version.

Reviewer 2: One thing that's not clear from this is where these modes are defined - if they are definable by each implementation, then we should say that explicitly here. (It would make sense to me that each implementation could define their own set of `mode`s.)

Author: It should be addressed now (see the other comment).

> omni-comprehensive one, as described [below](#multiple-reports-merge).

Reviewer 2: This seems like a fragment?

Author: yep, fixed 👍

### Implementation mode

The certification process runs against an implementation using a specific mode.
The final report will contain such a mode in the modes field. A "mode" is intended
to capture situations where a Gateway API implementation may have different features
and capabilities depending on how it is deployed (e.g. an implementation might be
deployed with an external, OR internal load balancer, and have different capabilities
depending on the mode chosen). The modes are implementation-specific, and no upstream

> mode is defined, except for `standard`, the default one used in case no mode is specified.

Author: I'm not 100% sure we should have a default mode. In case an implementation has a unique mode and is not interested in this API functionality, should we just leave the `modes` array empty instead of forcing a `standard` mode?

The more I think about it, the more convinced I get about leaving the array empty and not having a default `standard` mode.

@shaneutt @youngnick any thoughts on this regard?

Reviewer 2: Yeah, I think leaving the array empty by default is fine - but we just need to be clear that if there are multiple results with the same version, they must have different values for the `modes` array. (The same actually goes for having multiple versions with the same mode).

That means that we need to specify the `modes` array as optional with a defined behavior.

If we're going to do this though, we should put the compulsory field (`version`) first, and the optional field (`mode`) second in the docs.

For `mode`, we should use language like:

Implementations MAY have multiple entries for the same `version` if and only if each entry has a distinct `modes` array. Having two results with the same `version` and identical `modes` arrays is invalid.

Reviewer 3: If we go with my suggestion below, I'd recommend that mode is not a list and just a single `string`, and we default to the Gateway Class Name used in the test run as `mode` unless a different value is specified.

release time. The script hack/verify-all.sh will be updated to ensure the
generated `.go` file is up to date with the VERSION file.

### Multiple reports merge

> Every profile gives the possibility to specify multiple modes and
> implementation versions, but the testing process is related to a specific mode
> and a specific version. In order to produce reports with multiple modes and versions,

Reviewer: ```suggestion
In order to produce a single report which contains reports from multiple test
runs each testing against distinct _modes_,
```

> a proper flag `merge-report-with` can be set when running the certification process.

Reviewer: ```suggestion
the flag `merge-report-with` can be set when running the certification process.
```

> The flag value is taken, and if no file with such a name exists, it gets created

Reviewer: ```suggestion
The flag value is expected to be a file name. If no file with exists with the given name, it gets created
```

and the report is written therein. If the file already exists, the suite merges
the current report with the given one. By proceeding this way, it is possible to
iteratively run multiple times the certification suite with different setups
(modes and versions) and provide a unique report.

> 

Reviewer: I think it's a minor thing, but it might be helpful to add a default behavior of not overwriting data, and enable a flag:

```suggestion

> **Warning**: By default trying to merge a new report that contains a mode that already exists in the 
> existing report specified by `merge-report-with` will result in an error, to avoid automatically deciding
> which report is the one the user actually intended to provide and then deleting data. Implementors
> will need to manually remove a mode from an existing report if they wish to override, or they can use
> the `merge-report-overwrite-existing` flag to explicitly indicate this is what they want.
```

Author: Yeah, it makes sense!

release time. The script hack/verify-all.sh will be updated to ensure the
generated `.go` file is up to date with the VERSION file.

> ### Multiple reports merge

Reviewer 3: Merging seems like it could get pretty complex. If different `modes` have different supported features are we just going to take the intersection of them? What about failing tests?

For GKE, I think I'd rather have different files for each unique mode. In our case it would be per GatewayClass, and we already have [8 unique GatewayClasses that we support](https://cloud.google.com/kubernetes-engine/docs/how-to/gatewayclass-capabilities).

I think I'd rather take one step closer to [upstream k8s conformance reporting structure](https://github.com/cncf/k8s-conformance/tree/master/v1.28) here and end up with something like this:

```
- v1.0
  - gke
    - README.md
    - mode-1-report.yaml
    - mode-2-report.yaml
```

We could potentially take this one step further and pull some of the product specific information out of the report and into a separate `product.yaml` (eg: https://github.com/cncf/k8s-conformance/blob/master/v1.28/gke/PRODUCT.yaml).

I don't know, I'm probably missing some of the goals here, so let me know if these recommendations don't make sense.

Author: > Merging seems like it could get pretty complex. If different modes have different supported features are we just going to take the intersection of them? What about failing tests?

The idea is to have a section for each mode running a specific implementation version. If different modes have different supported features, we create two different sections, one per mode, along with the `supportedFeatures` and the results (successful, failed, skipped tests).

>For GKE, I think I'd rather have different files for each unique mode. In our case it would be per GatewayClass, and we already have [8 unique GatewayClasses that we support](https://cloud.google.com/kubernetes-engine/docs/how-to/gatewayclass-capabilities).
>
>I think I'd rather take one step closer to [upstream k8s conformance reporting structure](https://github.com/cncf/k8s-conformance/tree/master/v1.28) here and end up with something like this:
>
>```- v1.0
>  - gke
>    - README.md
>    - mode-1-report.yaml
>    - mode-2-report.yaml
>```
> We could potentially take this one step further and pull some of the product specific information out of the report and into a separate product.yaml (eg: https://github.com/cncf/k8s-conformance/blob/master/v1.28/gke/PRODUCT.yaml).

I get your point, and I agree that with the proposed solution we can end up having a gigantic `yaml` file. I like the approach of having a "summary report file" and a set of reports for each mode/version pair (or something similar).
I'll put something together that takes into account these considerations.

Author: @robscott it took some time, but eventually, I came up with a proposal very similar to what you suggested. Let me know what you think 😄

Reviewer 3: Thanks @mlavacca! Sorry for the delay on reviewing this!

> regarding conformance, e.t.c.). Optionally, it can be an email address or
> a support URL (e.g. Github new issue page).

The above report describes an implementation that just released `v1`, uses gateway
API `v0.8.0` `standard` channel, and has:

* `HTTP` `core` and `extended` support in `standard` mode;
* `HTTP` `core` and `extended` support in `enhanced` mode;

> * `TCP` `Core` support in `enhanced` mode.

Reviewer: From the perspective of this document, I don't feel like `enhanced` has enough meaning or context here.

Author: Do you have other words in mind that could suit better our needs here?

Reviewer 2: I put a note about this above, but if I'm right and the implementation can define the modes, then maybe a note here as well that says something like "remember that the mode values are arbitrary and chosen by the implementation" or something.

We could also make that clearer by making the names more clearly arbitrary - like calling the two modes `plain` and `with-the-lot` or something silly.

Author: > I put a note about this above, but if I'm right and the implementation can define the modes, then maybe a note here as well that says something like "remember that the mode values are arbitrary and chosen by the implementation" or something.

I agree that we need to specify this aspect; I added it [here](https://github.com/kubernetes-sigs/gateway-api/blob/342075e387f4b6534921be5b172dad734b705e7b/geps/gep-1709.md#implementation-mode).

as suggested by Nick, I changed the modes to `plain` and `with-the-lot`, to better highlight this aspect.

API version folders for which they have a supported implementation. The implementors
folder contains the following set of files:

* `README.md`: contains general information about the implementation and a table
  with some information, as shown below, to give users a quick overview of
  the implementation details.
* `<API-channel>-<implementation-version>-<mode-name>-report.yaml`: the report
  generated by the conformance test suite for the specific mode.

#### README.md structure

The README.md files SHOULD be structured in the following way (enrichment from
implementors is allowed, even though the minimum content is described below):

```md
# Acme

General information about the Acme/operator project

## Support Matrix

| API channel | Implementation version | Mode | Profiles | Extended Features | Report |
|-------------|------------------------|------|----------|-------------------|--------|
|             |                        |      |          |                   |        |
|             |                        |      |          |                   |        |
|             |                        |      |          |                   |        |

> 

Author: The goal of the PR is to define different modes and versions for a specific report. This aspect comes with a challenge: how to organize multiple reports for an implementation against the same Gateway API version? For example, with Gateway API, acme/operator may have many reports, with different operator versions and modes. I'd like to introduce here a way of organizing reports (originally proposed by @robscott in a [comment](https://github.com/kubernetes-sigs/gateway-api/pull/2391#discussion_r1327857784)), along with a table of contents. I think the table of contents is something beneficial at this point, but if we want to keep it for a subsequent GEP update, I have no problem at all 👍

> 

Reviewer: Alright, no that's fine let's leave it then. Thanks for the explanation :vulcan_salute:
